common/scpi.c

```diff
@@ -0,0 +1,217 @@
+#include <string.h>
+#include <util.h>
+#include <wallclock.h>
+#include <work.h>
+#include <platform/memory.h>
+#include <platform/time.h>
+
+#define SCPI_BUFFER_MAX  2
+
+#define SCPI_MEM_AREA(n) (((struct scpi_mem *)SCPI_MEM_BASE)[n])
+
+#define SCPI_TX_TIMEOUT  (100 * REFCLK_KHZ)     /* 100ms */
+
+struct scpi_mem {
+	struct scpi_msg tx_msg; /**< The reply to be sent to a client. */
+	struct scpi_msg rx_msg; /**< The request received from a client. */
+} __packed;
+
+struct scpi_buffer {
+	struct scpi_mem mem;    /**< Memory for the request/reply messages. */
+	uint8_t         client; /**< Client that should receive the reply. */
+	bool            busy;   /**< Flag telling if this buffer is in use. */
+};
+
+/** The message box device used for communication with SCPI clients. */
+static struct device *scpi_msgbox;
+
+/** Buffers used to hold messages while they are being processed. */
+static struct scpi_buffer scpi_buffers[SCPI_BUFFER_MAX];
+
+static struct scpi_buffer *
+scpi_alloc_buffer(uint8_t client)
+{
+	for (size_t i = 0; i < SCPI_BUFFER_MAX; ++i) {
+		if (!scpi_buffers[i].busy) {
+			struct scpi_buffer *buffer = &scpi_buffers[i];
+
+			/* Mark the buffer as being in use. */
+			buffer->busy   = true;
+			buffer->client = client;
+
+			return buffer;
+		}
+	}
+
+	return NULL;
+}
+
+/**
+ * Part 3 of SCPI message handling.
+ *
+ * Send a prepared message once the shared memory area is free (the AP has
+ * acknowledged our last message). Try hard to avoid dropping messages. If
+ * waiting for an acknowledgement times out, do other work and try again later.
+ */
+static void
+scpi_send_message(void *param)
+{
+	int err;
+	struct scpi_buffer *buffer = param;
+	uint8_t  client  = buffer->client;
+	uint64_t timeout = wallclock_read() + SCPI_TX_TIMEOUT;
+
+	/* Wait for any previous message to be acknowledged, with a timeout. */
+	while (msgbox_tx_pending(scpi_msgbox, buffer->client)) {
+		if (wallclock_read() >= timeout) {
+			warn("SCPI.%u: Channel busy, message dropped", client);
+			buffer->busy = false;
+			return;
+		}
+	}
+
+	/* Copy the prepared reply from the buffer to shared memory. */
+	memcpy(&SCPI_MEM_AREA(client).tx_msg, &buffer->mem.tx_msg,
+	       SCPI_HEADER_SIZE + buffer->mem.tx_msg.size);
+
+	/* Notify the client that the message has been sent. */
+	if ((err = msgbox_send_msg(scpi_msgbox, client, SCPI_VIRTUAL_CHANNEL)))
+		error("SCPI.%u: Error sending reply: %d", client, err);
+
+	/* Mark the buffer as no longer in use. */
+	buffer->busy = false;
+}
+
+/**
+ * Parts 1 & 2 of SCPI message handling (SCP-initiated messages).
+ */
+int
+scpi_create_message(uint8_t client, uint8_t command, uint32_t *payload,
+                    uint16_t payload_size)
+{
+	struct scpi_buffer *buffer = scpi_alloc_buffer(client);
+
+	/* Ensure there is space to put the created message. */
+	if (buffer == NULL)
+		return EAGAIN;
+
+	/* Create the message header. */
+	buffer->mem.tx_msg.command = command;
+	buffer->mem.tx_msg.sender  = 0;
```

Comment:
For readability, this should use the macro.

Reply:
What macro? Do you want me to create one?

Reply:
~the~ -> a. I think it makes sense to create one (or leave an inline comment); at line 114 this is not clear and requires extra reading.

Reply:
I added a macro